File: CodeCatch.js
// 실행 방법 ----------------------------------------
// 1. Nodejs가 반드시 설치가 되어 있어야 합니다!
// 2. CodeCatch.js 파일 위치에서 터미널을 열고 npm init -y 초기화
// 3. npm install ws 로 WebSocket을 설치
// 4. node CodeCatch.js 명령어로 실행
// -----------------------------------------------

const PinNumber = "________"; //여기에 핀 번호를 입력해주세요.

// 수정가능한 데이터 입니다. ----------
let attack = false; // boolean : false or true
let angle = 90.0; // float : 0~360.0
let type = 0; // int : 0,1,2,3
// ----------------------------

let CatchObj; //전달 받는 오브젝트 정보

// 수정 하시면 안됩니다! -------------------------------------------------------------------------------
const WebSocket = require("ws");
const socket = new WebSocket("ws://k7c106.p.ssafy.io:8080");

function run() {
  socket.onopen = function () {
    var Data = {
      action: "CodeInit",
      pinnumber: PinNumber,
    };
    socket.send(JSON.stringify(Data));
  };

  socket.onmessage = function (data) {
    var msg = JSON.parse(data.data.toString());

    //연결 성공
    if (msg.action === "ConnectSuccess") {
      console.log(msg.log);
    }
    //연결 실패시
    else if (msg.action === "ConnectFail") {
      console.log(msg.log);
      process.exit(1);
    }
    // 데이터를 전달 받음
    else if (msg.action === "exeData") {
      //데이터를 저장
      CatchObj = msg.catchobj;

      play(); //유저가 작성한 코드 수행

      var AttackInfo = {
        action: "codeData",
        pinnumber: PinNumber,
        attack: attack,
        angle: angle,
        type: type,
      };
      socket.send(JSON.stringify(AttackInfo));
    }
  };
}

run();
// --------------------------------------------------------------------------

function play() {
  // 여기서부터 코드를 작성 해주시면 됩니다. ------------------------------------------------------------
  for (let i = 1; i < CatchObj.length; ++i) {
    console.log(
      "위치 정보 : [" +
        CatchObj[i][0] +
        "," +
        CatchObj[i][1] +
        "] 타입 정보 : " +
        CatchObj[i][2]
    );
  }
  angle = 45;
  attack = true;
  type = 1;

  console.log("-------------------------------");

  // ----------------------------------------------------------------------------------------
}

File: CodeCatch.py
# 실행 방법 ----------------------------------------
# 1. pip install websockets : 웹 소켓 모듈을 선언합니다.
# 2. python CodeCatch.py로 실행
# -----------------------------------------------

PinNumber = "77751005"  # 화면에 나타난 PinNumber를 입력해 주세요.

import asyncio
import json
import websockets

# 수정가능한 데이터 입니다. -----------------
attack = False # Bool
angle = 0   # 0.0 ~ 360.0
type = 0     # int : 0,1,2,3
# -----------------------------------

async def Attack(websocket, catchobj):
    # 여기서부터 코드를 작성해주세요. --------------------------------------
    # catchobj는 현재 적의 위치 좌표가 들어간 배열입니다.
    # catchobj[i][0] -> 객체 x 위치 좌표
    # catchobj[i][1] -> 객체 y 위치 좌표
    # catchobj[i][2] -> 객체 타입 정보
    print(catchobj)

    attack = True
    angle = 45  
    type = 0   

    #---------------------------------------------------------------------------------

    # 아래는 건드리면 안 됩니다.----------------------------------------------------------------
    return {
                'action' : "codeData",
                'pinnumber' : PinNumber,
                'attack' : attack,
                'angle' : angle,
                'type' : type
            }

async def run(websocket):
    Data = {
        "action": "CodeInit",
        "pinnumber" : PinNumber
    }
    newData = json.dumps(Data)
    print(newData)
    await websocket.send(newData)
    msg = await websocket.recv()
    recvMsg = json.loads(msg)
    if recvMsg['action'] == 'ConnectSuccess':
        print("연결 성공")
    elif recvMsg['action'] == 'ConnectFail':
        print("연결 실패")
        exit()
    
    while 1:
        msg = await websocket.recv()
        recvMsg = json.loads(msg)
        if recvMsg['catchobj'] == [[]]:
            break
        if recvMsg['action'] == 'exeData':
            AttackInfo = await Attack(websocket, recvMsg['catchobj'])
            attackinfo = json.dumps(AttackInfo)
            await websocket.send(attackinfo)
async def connect():
  # 웹 소켓에 접속을 합니다.
  async with websockets.connect("ws://k7c106.p.ssafy.io:8080") as websocket:
    # 10번을 반복하면서 웹 소켓 서버로 메시지를 전송합니다.
    await run(websocket)
# 비동기로 서버에 접속한다.
asyncio.get_event_loop().run_until_complete(connect())
#---------------------------------------------------------------------------------